crates/compiler/mono/src/layout.rs

```diff
@@ -38,11 +38,397 @@ roc_error_macros::assert_sizeof_default!(Layout, 4 * 8);
+        self.raw_function_cache.push(Default::default());
+        CacheSnapshot {
+            layer: self.cache.len(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn rollback_to(&mut self, snapshot: CacheSnapshot) {
+        let CacheSnapshot { layer } = snapshot;
+
+        debug_assert_eq!(self.cache.len(), layer);
+        debug_assert_eq!(self.raw_function_cache.len(), layer);
+
+        self.cache.pop();
+        self.raw_function_cache.pop();
+    }
+
+    /// Invalidates the list of given root variables.
+    /// Usually called after unification, when merged variables with changed contents need to be
+    /// invalidated.
+    pub fn invalidate(&mut self, vars: impl IntoIterator<Item = Variable>) {
+        for var in vars.into_iter() {
+            for layer in self.cache.iter_mut().rev() {
+                layer.0.remove(&var);
+                roc_tracing::debug!(?var, "invalidating cached layout");
+            }
+            for layer in self.raw_function_cache.iter_mut().rev() {
+                layer.0.remove(&var);
+                roc_tracing::debug!(?var, "invalidating cached layout");
+            }
+        }
+    }
+
+    #[cfg(debug_assertions)]
+    pub fn statistics(&self) -> (CacheStatistics, CacheStatistics) {
+        (self.stats, self.raw_function_stats)
+    }
+}
+
+pub struct CacheSnapshot {
+    /// Index of the pushed layer
+    layer: usize,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct CacheCriteria {
+    /// Whether there is a naked recursion pointer in this layout, that doesn't pass through a
+    /// recursive structure.
+    has_naked_recursion_pointer: bool,
+    /// Whether this layout contains a recursive structure. If `Some`, contains the variable of the
+    /// recursion variable of that structure.
+    has_recursive_structure: Option<Variable>,
+}
+
+const CACHEABLE: CacheCriteria = CacheCriteria {
+    has_naked_recursion_pointer: false,
+    has_recursive_structure: None,
+};
+
+const NAKED_RECURSION_PTR: CacheCriteria = CacheCriteria {
+    has_naked_recursion_pointer: true,
+    has_recursive_structure: None,
+};
+
+impl CacheCriteria {
+    #[inline(always)]
+    fn is_cacheable(&self) -> bool {
+        // Can't cache if there a naked recursion pointer that isn't covered by a recursive layout.
+        !self.has_naked_recursion_pointer
+    }
+
+    /// Makes `self` cacheable iff self and other are cacheable.
+    #[inline(always)]
+    fn and(&mut self, other: Self) {
+        self.has_naked_recursion_pointer =
+            self.has_naked_recursion_pointer || other.has_naked_recursion_pointer;
+        // TODO: can these ever conflict?
+        self.has_recursive_structure = self
+            .has_recursive_structure
+            .or(other.has_recursive_structure);
+    }
+
+    #[inline(always)]
+    fn pass_through_recursive_union(&mut self, recursion_var: Variable) {
+        self.has_naked_recursion_pointer = false;
+        self.has_recursive_structure = Some(recursion_var);
+    }
```

Comment:
won't the recursion pointer be inevitable at this point?

Reply:
Yes, but the key is, if it has a recursive structure, there is no naked recursion pointer. The `has_naked_recursion_pointer` is to avoid caching layouts like `List (RecursivePointer)` rather than `List (NullableUnwrapped ...)`